escrow/src/main.sw

```diff
@@ -1,257 +1,346 @@
 contract;
 
+// TODO: commented-out code is there because the SDK does not support Vec yet so an array is used
+
+// Our library dependencies
+dep data_structures;
+dep errors;
+dep events;
+dep interface;
+
+use data_structures::{Arbiter, Asset, Buyer, EscrowInfo, Seller, State};
+use errors::{
+    ArbiterInputError,
+    AssetInputError,
+    DeadlineInputError,
+    DepositError,
+    StateError,
+    UserError,
+    UserInputError,
+};
+
+use events::{
+    AcceptedArbiterEvent,
+    CreatedEscrowEvent,
+    DepositEvent,
+    DisputeEvent,
+    PaymentTakenEvent,
+    ProposedArbiterEvent,
+    ResolvedDisputeEvent,
+    ReturnedDepositEvent,
+    TransferredToSellerEvent,
+};
+
+use interface::Escrow;
 use std::{
-    address::Address,
     assert::require,
-    chain::auth::{AuthError, msg_sender},
-    context::{call_frames::{contract_id, msg_asset_id}, msg_amount, this_balance},
+    block::height,
+    chain::auth::msg_sender,
+    context::{call_frames::msg_asset_id, msg_amount},
     contract_id::ContractId,
     identity::Identity,
-    result::*,
-    revert::revert,
-    token::transfer_to_output,
+    logging::log,
+    option::Option,
+    result::Result,
+    storage::StorageMap,
+    token::transfer,
 };
 
-abi Escrow {
-    #[storage(read, write)]fn constructor(buyer: Address, seller: Address, asset: ContractId, asset_amount: u64) -> bool;
+storage {
+    /// Used as a temporary variable for containing a change, proposed by the seller, to the arbiter
+    /// Map(ID => Info)
+    arbiter_proposal: StorageMap<u64,
+    Option<Arbiter>> = StorageMap {
+    },
 
-    #[storage(read, write)]fn deposit() -> bool;
+    /// Information describing an escrow created via create_escrow()
+    /// Map(ID => Info)
+    escrows: StorageMap<u64,
+    EscrowInfo> = StorageMap {
+    },
 
-    #[storage(read, write)]fn approve() -> bool;
+    /// Number of created escrows
+    /// Used as an identifier for O(1) look-up in mappings
+    escrow_count: u64 = 0,
+}
 
-    #[storage(read, write)]fn withdraw() -> bool;
+impl Escrow for Contract {
+    #[storage(read, write)]fn accept_arbiter(identifier: u64) {
```

Comment:
No need since the escrow can only have 1 value set at a time so the user should be able to see the proposition and all they have to do is accept it otherwise I'd need to implement some juggling of any number of arbiters being proposed at a time and then handle all of those fund transfers.

Reply:
The one problem that I see is the seller creating a new arbiter proposal a block before the buyer calls `accept_arbiter` not giving the front end time to update. This may be a stretch but a possible attack vector and something worth considering.

Reply:
I don't think that can be solved on-chain. We don't know block times so I cannot select an arbitrary number of blocks and that may also mess up the deadline. If I did know the times that wouldn't help much anyway imo. This is an edge case as it would need to occur right before the acceptance which is unlikely to happen or front running must be done and the chance of some node allowing this specifically is also unlikely so it's possible but very slim.